### ISSUE TITLE
Conversion to Verona MLIR dialect

### ISSUE THREAD

**Issue by rengolin** [microsoft/verona#107]
With both dialect and type system designed, we can start coding the conversion infrastructure from opaque MLIR to Verona dialect. This task will probably need to occur in parallel with #105, and after #104 and #106.

We should have a series of partial conversions, with or without type inference, and finish with a pure Verona dialect (with possible embedded std/loop dialects) graph.

Acceptance criteria:
* Write all partial conversions from opaque MLIR to Verona dialect
* Finish with a full conversion and validate that the dialect is pure and consistent
* Write tests for all known cases of the syntax/semantics and some larger programs

Not included:
* Lowering to LLVM dialect
* Optimisation passes other than required to validate

**Comment by rengolin** [comment]
Stub dialect inside the Verona tree with LIT test working at https://github.com/rengolin/verona/tree/mlir-ast.

So far, it reads opaque MLIR and spits out broken MLIR, but it's just a stub for the infrastructure, not the actual code logic.

The AST -> MLIR bridge will be built once we have the parser in-tree (#111), and it's also when we'll be able to start working on the dialect proper (#106).

**Comment by rengolin** [comment]
Putting this on hold until we have an actual AST->MLIR conversion (#117)

**Comment by rengolin** [comment]
Ok, now it's time to start lowering into the Verona dialect. The first task is #203.

**Comment by rengolin** [comment]
We're moving the high level IR to a different design, only having the low-level parts in MLIR, so all of the designs in this issue are being deprecated in favour of a more modular approach.
